### PR TITLE
chore(eslint): Setup eslint and prettier

### PR DESCRIPTION
**Pull request by hbuchel**
#### Description of changes:

This PR sets up eslint and fixes the existing linting errors.

- It follows Next.js' [recommended setup](https://nextjs.org/docs/pages/building-your-application/configuring/eslint) using the "strict" option
- `eslint-config-next` already configures `eslint-plugin-react` and `eslint-plugin-react-hooks`
- Adds `eslint-plugin-prettier` so we can enable prettier
- Added a rule to treat `react-hooks/exhaustive-deps` as an error instead of a warning
- MDX files are currently ignored for linting/prettier in this PR (those need a bit more testing and additional plugins installed)

Running `yarn lint` on this PR should not produce any eslint errors.

[Preview](https://setupeslint.d2bfwhpcsj9awv.amplifyapp.com/)

**Prereqs**
- ✅ ~Merge https://github.com/aws-amplify/docs/pull/6566 first~ 

**After merging this PR, the following can be merged** These have been split up so that the PRs are a little easier to digest and review.
- Add eslint/typescript: https://github.com/aws-amplify/docs/pull/6577
- Run prettier against all files: https://github.com/aws-amplify/docs/pull/6580

**Questions/Comments:**
- Any performance concerns for making `awshome_s_code` script async? eslint-config-next complains about 3rd party scripts loaded asynchronously. Using `defer` throws the following error on tracked events (screenshot below) 
    - **Answer:** These have eslint-ignore comments because metrics team said they need to be loaded synchronously
Screenshot 2023-11-29 at 5 54 56 PM

- lint-staged latest (15) is incompatible with node16 and requires at least 18.12. I downgraded to lint-staged@^14 for now.

```
error lint-staged@15.1.0: The engine "node" is incompatible with this module. Expected version ">=18.12.0". Got "16.20.2"
error Found incompatible module.
```

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [ ] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [ ] JS
- [ ] iOS
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [ ] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://link.com)` 
            HTML: `<a href="https://link.com">link</a>`_

### When this PR is ready to merge, please check the box below
- [ ] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

**Review by hbuchel — COMMENTED** [review]
Notes about some of the changes:

**Review by timngyn — DISMISSED** [review]

File: src/data/platforms.ts [hunk 1]
```diff
@@ -1,23 +1,33 @@
...
-  'android' | 
-  'angular' | 
-  'flutter' | 
-  'javascript' | 
-  'nextjs' | 
-  'react' | 
-  'react-native' | 
-  'swift' | 
-  'vue';
+export type Platform =
```
Comment (reesscot) [inline src/data/platforms.ts:1]: Could be simplified to this:
```
export type Platform =
  | 'android'
  | 'flutter'
  | 'swift'
  | JSPlatform
```

File: src/utils/useRouteFinder.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import flatDirectory from 'src/directory/flatDirectory.json';
+import { useRouter } from 'next/router';
+import { Platform } from '@/data/platforms';
+
+export const useRouteFinder = (platform: Platform, isPrev: boolean) => {
```
Comment (reesscot) [inline src/utils/useRouteFinder.ts:5]: Can we follow up with a unit test?

**Review by sowmyadev-kota — APPROVED** [review]

File: src/pages/404.tsx [hunk 3]
```diff
@@ -29,8 +29,9 @@ export default function Custom404() {
     <Flex className="four-oh-four">
       <Heading level={1}>404</Heading>
       <Text>
-        Apologies––we can't seem to find the page for which you're looking. If
-        this is a mistake, please file an issue and we'll fix it ASAP.
+        Apologies &#8212; we can&apos;t seem to find the page for which
+        you&apos;re looking. If this is a mistake, please file an issue and
+        we&apos;ll fix it ASAP.
```
[the remark this answers is not in the document]
Comment (hbuchel) [reply src/pages/404.tsx:34]: We chatted offline but adding the reply here: eslint complains about the single quotes in code and recommends using the html entity `&apos;` or a curly quote. This should only affect our code files and not MDX files for authors.

File: src/utils/useRouteFinder.ts (hunk 2 above)
Comment (hbuchel) [reply src/utils/useRouteFinder.ts:5]: Yes, added a task to capture that.

**Review by reesscot — APPROVED** [review]
LGTM

**Review by timngyn — APPROVED** [review]

**Pull request merged**